stdlib/camlinternalLazy.ml

```diff
@@ -42,7 +42,7 @@ external update_to_forward : Obj.t -> unit =
 
 (* Assumes [blk] is a block with tag forcing *)
 let do_force_block blk =
-  let b = Obj.repr blk in
+  let b = Obj.repr (Sys.opaque_identity blk) in
```

Comment:
This use of `Sys.opaque_identity` is not explained by a comment. Could you add a comment to explain what is going on?

In fact, the comment you added in 0f30379ad94f81360586465a840107a76a5b486b has disappeared from trunk, probably dropped during the Multicore rebases. Could you add it back? (I think it makes sense to add the two comments at the same time, they will be more consistent with each other this way.)

Reply:
Is this the right place to call `Sys.opaque_identity`? (And if it is, why isn't `do_force_val_block` handled in the same way?) My understanding is that this function is consistent with the idea that `blk` is a mutable block, and that it is only called on values that do respect this assumption -- same thing with `force_gen_lazy_block`, which would have as precondition that the input is a mutable block, as it unconditionally calls `update_to_forcing`.

Where is the actual move from "possibly immutable" to "certainly mutable" going on? I think it takes place within the code of `force_gen`, so this is where `Sys.opaque_identity` should be placed. But there is one already. Maybe we should add a Popaque primitive in `inline_lazy_force_*` instead of here?

Reply:
> Where is the actual move from "possibly immutable" to "certainly mutable" going on? I think it takes place within the code of `force_gen`, so this is where `Sys.opaque_identity` should be placed. But there is one already. Maybe we should add a Popaque primitive in `inline_lazy_force_*` instead of here?

That's a good idea. I'll push an update with that solution.